### ISSUE TITLE
New rule: RedundantIfWithCheck/Require

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#4273]
## Expected Behavior of the rule
```kotlin
if (prices.isEmpty()) {
    check(prices.isNotEmpty()) { "Prices list after selection cannot be empty!" }
}
```
The hard thing is coming up with all the potential variations of this (e.g. `if (isNotEmpty) check(isEmpty)`, `if(!isEmpty) check(isEmpty)`, and `if(anything) check(false)`.

## Context
Reduce redundant complex code

**Comment by marschwar** [comment]
As you already mentioned yourself, this rule would be very hard (if even possible) to implement and I am not sure that detekt should  do such deep code analysis. On top of having to identify the negated functions, we would additionally have to guarantee that the result of the expression cannot change. Maybe this is possible as part of the kotlin compiler pipeline?

Are you talking about inverse functions in general or just a few specific ones.

**Comment by TWiStErRob** [comment]
I was thinking best effort hand maintained negations, the ones listed and a few more. At least cover stdlib and a few special cases, e.g. a local variable negated.

**Comment by marschwar** [comment]
To be quite honest, I feel the benefit of such a rule would be very slim compared to the high chance of false positives.

**Comment by BraisGabin** [comment]
> Maybe this is possible as part of the kotlin compiler pipeline?

I was thinking the same.

It would be a really nice rule but I don't know why the compiler doesn't do it directly. This is exactly the same as smart cast. Is there any issue in kotlin's issue tracker where intellij talks about why this is not implemented? Other option would be to check how the Compiler does the smart cast and see if we can adapt that code.
